sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/OptimizeLocalShuffleReader.scala

```diff
@@ -140,9 +140,9 @@ object OptimizeLocalShuffleReader {
 
   def canUseLocalShuffleReader(plan: SparkPlan): Boolean = plan match {
     case s: ShuffleQueryStageExec =>
-      s.shuffle.canChangeNumPartitions
+      s.shuffle.canChangeNumPartitions && s.mapStats.isDefined
```

Comment:
This skips 0 partitions case, right?

Reply:
yea, otherwise we will hit
```
val splitPoints = if (numMappers == 0) {
  Seq.empty
} else ...
```
which creates a local reader with 0 partitions.

Reply:
We should be able to turn the `if` into `assert`, but I'd like to only do it in master to be safe.